docs/signing-tool-spec.md

```diff
@@ -0,0 +1,129 @@
+
+## Challenges
+
+There are a few challenges around code signing:
+
+### Local Certificates
+
+Today the code signing tooling in the Windows and .NET SDK uses PFX (public/private certificate key pair files or the local certificate store for obtaining certificates). There risks to this approach:
+
+- PFX files are targets in data breaches; their passwords can be cracked
+- Certificates in a local store can be used by any app/malware
+- There’s no revocation mechanism for the users’ access to the certificate; they always have it
+- No auditing of signing operations possible
+- EV code signing certificates aren’t easily supported as they require FIPS 140-2 hardware devices with drivers
+
+In May 2022, the CA/B forum updated its baseline requirements to require HSMs so local certificates will no longer be issued for publically trusted roots. The only support the current signing tools have for this scenario is via CSP/KSP drivers provided by some HSM vendors and those do not work well for cloud-based build agents. The current tools would need new investment to support different backends.
+
+### Orchestration
+
+An application/library package typically contains multiple assets that need to be signed. For example, a NuGet package (`.nupkg` file) contains `.dll` files that also must be signed. An ClickOnce or MSIX package also contains `.dll` or `.exe` that need to be signed. A `.vsix` file can contain `.dll` files, and `.nupkg` file that must be signed. These files need to be signed "inside out" to ensure the proper sequence. That is, a `.vsix` containing a `.nupkg` needs to extract the inner `.nupkg` to sign the contained `.dll` files, then sign the `.nupkg` and any other `.dll` files, then repack and sign the `.vsix`. Other types, like ClickOnce and MSIX may contain manifest files that also must be updated during these operations.
+
+With several different code signing tools required, each with its own command line parameters and defaults. The process of code signing is error-prone and hard to get right. The signing tool addresses these challanges by unifying the interface into a single set of options.
+
+## Plan
+
+Create a modern signing tool to eventually replace the existing tools. The tool will handle all of our first party signing formats, orchestrate signing files in the right order, and have extensibility to support multiple signing providers. As our customers use a variety of clouds and HSMs, the extensibility will enable us to meet customers needs wherever they store their certificates.
```

Comment:
While having this tool replace existing tools is a good north star, I think we should hold-off on mentioning it until the tool gains ground with customers.